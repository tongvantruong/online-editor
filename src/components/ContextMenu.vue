<template>
  <ul v-if="visible" class="context-menu" :style="{ top: y + 'px', left: x + 'px' }">
    <li v-for="(act, index) in actions" :key="index" class="context-menu-item" @click="act.action()">
      {{ act.icon }}<span class="context-menu-label">{{ act.label }}</span>
    </li>
  </ul>
</template>

<script lang="ts" setup>
import { PropType } from 'vue'

export type MenuItem = {
  icon: string
  label: string
  action: () => void
}

const { visible, x, y, actions } = defineProps({
  visible: {
    type: Boolean,
    default: false,
  },
  x: {
    type: Number,
    default: 0,
  },
  y: {
    type: Number,
    default: 0,
  },
  actions: {
    type: Array as PropType<MenuItem[]>,
    default: [],
  },
})
</script>

<style scoped>
.context-menu {
  position: absolute;
  z-index: 1000;
  background-color: white;
  border: 1px solid #ccc;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 4px 0;
  min-width: 150px;
  list-style: none;
}

.context-menu-item {
  padding: 8px 12px;
  cursor: pointer;
}

.context-menu-item:hover {
  background-color: #f0f0f0;
}

.context-menu-label {
  margin-left: 8px;
}
</style>
