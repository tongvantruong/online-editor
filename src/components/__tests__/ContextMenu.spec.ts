import { mount } from '@vue/test-utils'
import ContextMenu, { MenuItem } from '../ContextMenu.vue'
import { describe, expect, it, vi } from 'vitest'

const visible = true
const mockDeleteAction = vi.fn()
const mockDuplicateAction = vi.fn()
const actions: MenuItem[] = [
  {
    icon: '❌',
    label: 'Delete',
    action: mockDeleteAction,
  },
  {
    icon: '📄',
    label: 'Duplicate',
    action: mockDuplicateAction,
  },
]

describe('ContextMenu', () => {
  it('renders Delete icon and label', () => {
    const wrapper = mount(ContextMenu, {
      props: { visible, actions },
    })
    const deleteItem = wrapper.findAll('.context-menu-item')[0]
    expect(deleteItem.text()).toBe(`${actions[0].icon}${actions[0].label}`)
  })
  it('renders Duplicate icon and label', () => {
    const wrapper = mount(ContextMenu, {
      props: { visible, actions },
    })
    const duplicateItem = wrapper.findAll('.context-menu-item')[1]
    expect(duplicateItem.text()).toBe(`${actions[1].icon}${actions[1].label}`)
  })
  it('calls Delete action when clicked Delete', () => {
    const wrapper = mount(ContextMenu, {
      props: { visible, actions },
    })
    wrapper.findAll('.context-menu-item')[0].trigger('click')
    expect(mockDeleteAction).toHaveBeenCalledOnce()
  })
  it('calls Duplicate action when clicked Duplicate', () => {
    const wrapper = mount(ContextMenu, {
      props: { visible, actions },
    })
    wrapper.findAll('.context-menu-item')[1].trigger('click')
    expect(mockDuplicateAction).toHaveBeenCalledOnce()
  })
})
