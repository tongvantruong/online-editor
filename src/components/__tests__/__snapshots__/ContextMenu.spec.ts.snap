// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContextMenu > matches snapshot 1`] = `
"<ul data-v-cbb365ae="" class="context-menu" style="top: 0px; left: 0px;">
  <li data-v-cbb365ae="" class="context-menu-item">❌<span data-v-cbb365ae="" class="context-menu-label">Delete</span></li>
  <li data-v-cbb365ae="" class="context-menu-item">📄<span data-v-cbb365ae="" class="context-menu-label">Duplicate</span></li>
</ul>"
`;

exports[`ContextMenu > matches snapshot 2`] = `
"<ul data-v-cbb365ae="" class="context-menu" style="top: 0px; left: 0px;">
  <li data-v-cbb365ae="" class="context-menu-item">❌<span data-v-cbb365ae="" class="context-menu-label">Delete</span></li>
  <li data-v-cbb365ae="" class="context-menu-item">📄<span data-v-cbb365ae="" class="context-menu-label">Duplicate</span></li>
</ul>"
`;
